### ISSUE TITLE
Support for seekable zstd format with Jumptable

### ISSUE THREAD

**Issue by tanantharaman** [martinellimarco/libzstd-seek#1]
I noticed you had a "TODO" about this issue in zstd-seek.c and I tested the solution on a 98TB backup using ratarmount and it eliminated over 24 hours regenerating the JumpTable during the mount. The solution is based on the code in zstd/contrib/seekable_format/zstdseek_decompress.c

I am new to git, so I don't know the right way to submit a change, so I am posting the C-code here : feel free to use it or adapt it(right after your "TODO" comment) :

// Adapted from zstd/contrib/seekable_format/zstdseek_decompress.c
#define ZSTD_seekTableFooterSize 9
#define ZSTD_SEEKABLE_MAGICNUMBER 0x8F92EAB1
#define ZSTD_SKIPPABLEHEADERSIZE    8
#define ZSTD_MAGIC_SKIPPABLE_START  0x184D2A50   
if(1) {
        void *buff = sctx->buff;
        size_t size = sctx->size;
        void *footer = buff + (size - ZSTD_seekTableFooterSize);
        unsigned magicnumber = *((unsigned *)(footer + 5));

        if(magicnumber == ZSTD_SEEKABLE_MAGICNUMBER){
            unsigned char sfd = *((unsigned char*)(footer + 4));
            unsigned checksumFlag = sfd >> 7;

            /* check reserved bits */
            if ((sfd >> 2) & 0x1f) {
               DEBUG("last frame checksumFlag= %x: Bits 3-7 should be zero\n",(unsigned int)sfd);
	       return -1;
            }
	    unsigned const numFrames = *((unsigned *)footer);
	    unsigned const sizePerEntry = 8 + (checksumFlag ? 4 : 0);
	    unsigned const tableSize = sizePerEntry * numFrames;
            unsigned const frameSize = tableSize + ZSTD_seekTableFooterSize + ZSTD_SKIPPABLEHEADERSIZE;
	    
            void *frame = buff + (size - frameSize);
	    unsigned skippableHeader = *((unsigned *)frame);
	    if(skippableHeader != (ZSTD_MAGIC_SKIPPABLE_START | 0xE)){
                 DEBUG("last frame Header = %u does not match magic number %u\n",skippableHeader, (ZSTD_MAGIC_SKIPPABLE_START | 0xE));
                 return -1;
            }

            unsigned FrameSize = *((unsigned *)(frame + 4));
            if(FrameSize + ZSTD_SKIPPABLEHEADERSIZE != frameSize){
                DEBUG("last frame size = %u does not match expected size = %u\n", FrameSize + ZSTD_SKIPPABLEHEADERSIZE, frameSize);
                return -1;
            }

            void *table = frame + ZSTD_SKIPPABLEHEADERSIZE;
            for(unsigned i = 0; i < numFrames; i++){
	        unsigned cOffset = *((unsigned *)(table + (i * sizePerEntry)));
	        unsigned dOffset = *((unsigned *)(table + (i * sizePerEntry) + 4));
	        ZSTDSeek_addJumpTableRecord(sctx->jt, cOffsetCum, dOffsetCum);
            }

            sctx->jumpTableFullyInitialized = 1;
            return 0;
	}
    }

**Comment by martinellimarco** [comment]
Hi there, thanks for the contribution.

The git way of doing this is to fork the project and then make a pull request, but don't worry, the code above is enough.

Il will work on this project again around Christmas, I have a few improvements here and in the companion project t2sz. Unfortunately my daily job is keeping me too busy :(

I'm intrigued, what have you used to generate a 98TB .tar.zstd backup in seekable format? I will need a way to encode some test files.

I have a question about `ZSTD_MAGIC_SKIPPABLE_START`. Why is it defined ad 0x184D2A50 and later you add 0xE with an bitwise OR to get 0x184D2A5E ? Can't you define it as 0x184D2A5E from the start? Am I missing something?

According to the [standard](https://github.com/facebook/zstd/blob/dev/contrib/seekable_format/zstd_seekable_compression_format.md) it's 0x184D2A5E.

**Comment by tanantharaman** [comment]
The 98TB is the uncompressed tar size, the .tar.zstd (using compression level 6) was around 30TB. Since I was using ext4, this meant the tar.zstd was actually broken into 2 pieces using "split -b 16TB", so I first had to figure out a way to present it to ratarmount as a single file, since your library uses mmap() to map the entire file : I used concatfs (https://github.com/schlaile/concatfs)), to create a FUSE mounted file of 30TB (it allows any file size up to 2^64 bytes, defined by a "cat" list of real files).

As for the ZSTD_MAGIC_SKIPPABLE_START : the comment in zstd/contrib/seekable_format/zstdseek_decompress.c states that the ZSTD format for skippable frame allows any of 15 possible values in the MAGIC number, and the seekable_format chose the one ending in 0xE, to identify the start of the JumpTable. I just followed their code, which I assume is trying to make clear that they picked one specific possible value out of 15 possible values under the ZSTD format for skippable frames.

BTW, I also modified ratarmount to allow the tarfile to be streamed from stdin (using a --stdin TARSIZE option), so the tar sqlite index can be created during the backup (using pipe "tee" to send one copy of the tar output stream to ratarmount and another copy to the seekable zstd compressor). The ratarmount then exits when the index is complete and prompts to be rerun the normal way : during the 2nd run  of ratarmount your library then computes the zstd frame JumpTable : with the above modification, the mount completes in a couple seconds for the 98TB backup volume! The original tar backup seems to take no more time than normal (around 12 days, limited by the GNU tar HDD read speed).
